process/transaction/shardProcess.go

```diff
@@ -633,34 +637,68 @@ func (txProc *txProcessor) addFeeAndValueToDest(acntDst state.UserAccountHandler
 	}
-	if tx.GetValue().Cmp(big.NewInt(0)) != 0 {
-		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3ZeroVal)
+	if check.IfNil(relayerAcnt) {
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrNilRelayerAccount)
+	}
+	err := txProc.relayedTxV3Processor.CheckRelayedTx(tx)
+	if err != nil {
+		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, err)
 	}
 
-	userTx := tx.GetInnerTransaction()
-	if !bytes.Equal(tx.RcvAddr, userTx.SndAddr) {
-		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3BeneficiaryDoesNotMatchReceiver)
+	// process fees on both relayer and sender
+	relayerFee, totalFee := txProc.relayedTxV3Processor.ComputeRelayedTxFees(tx)
+	err = txProc.processTxAtRelayer(relayerAcnt, totalFee, relayerFee, tx)
+	if err != nil {
+		return 0, err
 	}
-	if len(userTx.RelayerAddr) == 0 {
-		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3EmptyRelayer)
+
+	innerTxs := tx.GetInnerTransactions()
+
+	var innerTxRetCode vmcommon.ReturnCode
+	var innerTxErr error
+	executedUserTxs := make([]*transaction.Transaction, 0)
+	for _, innerTx := range innerTxs {
+		innerTxRetCode, innerTxErr = txProc.finishExecutionOfInnerTx(tx, innerTx)
+		if innerTxErr != nil || innerTxRetCode != vmcommon.Ok {
+			continue
+		}
+
+		executedUserTxs = append(executedUserTxs, innerTx)
 	}
-	if !bytes.Equal(userTx.RelayerAddr, tx.SndAddr) {
-		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3RelayerMismatch)
+
+	allUserTxsSucceeded := len(executedUserTxs) == len(innerTxs) && innerTxErr == nil && innerTxRetCode == vmcommon.Ok
+	if !allUserTxsSucceeded {
+		log.Trace("failed to execute all inner transactions", "total", len(innerTxs), "executed transactions", len(executedUserTxs))
 	}
-	if tx.GasPrice != userTx.GasPrice {
-		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedV3GasPriceMismatch)
+
+	return vmcommon.Ok, nil
+}
+
+func (txProc *txProcessor) finishExecutionOfInnerTx(
+	tx *transaction.Transaction,
+	innerTx *transaction.Transaction,
+) (vmcommon.ReturnCode, error) {
+	acntSnd, err := txProc.getAccountFromAddress(innerTx.SndAddr)
+	if err != nil {
+		return vmcommon.UserError, err
 	}
-	remainingGasLimit := tx.GasLimit - txProc.economicsFee.ComputeGasLimit(tx)
-	if userTx.GasLimit != remainingGasLimit {
-		return vmcommon.UserError, txProc.executingFailedTransaction(tx, relayerAcnt, process.ErrRelayedTxV3GasLimitMismatch)
+
+	if check.IfNil(acntSnd) {
+		return vmcommon.UserError, process.ErrRelayedTxV3SenderShardMismatch
 	}
 
-	return txProc.finishExecutionOfRelayedTx(relayerAcnt, acntDst, tx, userTx)
+	txFee := txProc.computeTxFee(innerTx)
+	err = txProc.addFeeAndValueToDest(acntSnd, tx, txFee)
```

Comment:
you can change the addFeeAndValueToDest function and have 2 bigInts to be received there. it is much more clear. and here you could have big.newint(0), txFee.